### ISSUE TITLE
eslint-config-next@12.0.5 gives "Type '() => any' is not a constructor function type" error in _Document.tsx

### ISSUE THREAD

**Issue by JamesSingleton** [vercel/next.js#32110]
### What version of Next.js are you using?

12.0.5

### What version of Node.js are you using?

14.18.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

After upgrading to both `next@12.0.5` and `eslint-config-next@12.0.5`, my `_Document.tsx` now throws an eslint error `Type '() => any' is not a constructor function type`. I have `typescript@^4.5.2` installed.

## _Doucment.tsx
```tsx
import Document, { Head, Html, Main, NextScript } from 'next/document'

class MyDocument extends Document {
  render() {
    return (
      <Html lang="en-US" className="h-full bg-gray-100">
        <Head />
        <body className="loading h-full">
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```

## .eslintrc.json

```json
{
  "extends": ["next/core-web-vitals", "prettier"],
  "rules": {
    "react/no-unescaped-entities": "off"
  }
}
```

### Expected Behavior

This was not breaking on 12.0.4 and should not break on 12.0.5

### To Reproduce

Create a Next.js project with the example code I provided above.

**Comment by indiejoseph** [comment]
+1

**Comment by emt-mocha** [comment]
upgraded to 12.0.5 and facing the same issue during build time. I am using: 
Typescript = 4.4.4
Eslint = 7.32.0
eslint-config-next = 12.0.5
nodejs = v16.13.0

**Comment by TimNZ** [comment]
I respect and appreciate the awesome work of the Next js team,
but I am perplexed how this got through testing for release.

The first line has been added to document.d.ts.
You can safely replace the contents with the 12.0.4 ver

12.0.5
```
import Document from './dist/shared/lib/runtime-config'
export * from './dist/pages/_document'
export default Document
```

12.0.4
```
export * from './dist/pages/_document'
export { default } from './dist/pages/_document'
```

**Comment by raulfdm** [comment]
It seems the usage is different now... or at least that's how the types are defined:
https://github.com/vercel/next.js/pull/31560/files#diff-82360298eb44f75969189468a71ab683227c0848c9b4fecbeae1321a7c1680b3R46

The error go away if I do:

```tsx
export default class MyDocument extends Document() {
 // ...rest
}
```

**Comment by TimNZ** [comment]
@raulfdm 
https://github.com/vercel/next.js/issues/32110#issuecomment-985977369

**Comment by timneutkens** [comment]
Hey! Thanks for the report and investigation. I've opened a PR here: #32117. The reason this got through to a release is that there was no custom _document.tsx / _app.tsx in the TypeScript test suites, I've added those in #32117.

**Comment by timneutkens** [comment]
Published 12.0.6: https://github.com/vercel/next.js/releases/tag/v12.0.6

It'll take a bit for it to be available on npm 🙏

**Comment by captain-woof** [comment]
> It seems the usage is different now... or at least that's how the types are defined: https://github.com/vercel/next.js/pull/31560/files#diff-82360298eb44f75969189468a71ab683227c0848c9b4fecbeae1321a7c1680b3R46
> 
> The error go away if I do:
> 
> ```tsx
> export default class MyDocument extends Document() {
>  // ...rest
> }
> ```
> 
> But if that's the case, it seems a breaking change. Who customize `_document` will need to change to that.
> 
> ## Update
> I run a build with that syntax and got an error:
> 
> ```
> TypeError: Class constructor Document cannot be invoked without 'new'
> ```
> 
> So I think the interface is incorrect
> 
> ## Workaround
> Quick and dirty:
> 
> ```tsx
> import Document, { Head, Html, Main, NextScript } from 'next/document';
> 
> const CustomDocument = Document as any
> 
> export default class MyDocument extends CustomDocument {
>   // ... rest
> }
> ```

Thanks. This helped!!!

**Comment by captain-woof** [comment]
> Published 12.0.6: https://github.com/vercel/next.js/releases/tag/v12.0.6
> 
> It'll take a bit for it to be available on npm pray

Awesome!

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
